xbmc/pvr/channels/PVRChannelProvider.cpp

```diff
@@ -0,0 +1,544 @@
+#include "utils/StringUtils.h"
+#include "utils/Variant.h"
+#include "utils/log.h"
+
+#include <memory>
+#include <string>
+
+using namespace PVR;
+
+CPVRChannelProvider::CPVRChannelProvider(int iUniqueId, int iClientId)
+  : m_iProviderId(PVR_PROVIDER_INVALID_UID),
+    m_iUniqueId(iUniqueId),
+    m_iClientId(iClientId),
+    m_providerType(PVR_CHANNEL_PROVIDER_TYPE_UNKNOWN),
+    m_bIsClientProvider(false)
+{
+  UpdatePath();
+}
+
+CPVRChannelProvider::CPVRChannelProvider(const PVR_CHANNEL_PROVIDER& provider, int iClientId)
+  : m_iProviderId(PVR_PROVIDER_INVALID_UID),
+    m_iUniqueId(provider.iUniqueId),
+    m_iClientId(iClientId),
+    m_strProviderName(provider.strProviderName),
+    m_strClientProviderName(provider.strProviderName),
+    m_providerType(provider.providerType),
+    m_clientProviderType(provider.providerType),
+    m_strIconPath(provider.strIconPath),
+    m_strClientIconPath(provider.strIconPath),
+    m_strCountry(provider.strCountry),
+    m_strClientCountry(provider.strCountry),
+    m_strLanguage(provider.strLanguage),
+    m_strClientLanguage(provider.strLanguage),
+    m_bIsClientProvider(false)
+{
+  UpdatePath();
+}
+
+CPVRChannelProvider::CPVRChannelProvider(const std::shared_ptr<CPVRClient>& client, int iClientId)
+  : m_iProviderId(PVR_PROVIDER_INVALID_UID),
+    m_iUniqueId(PVR_PROVIDER_ADDON_UID),
+    m_iClientId(iClientId),
+    m_strProviderName(client->Name()),
+    m_strClientProviderName(client->Name()),
+    m_providerType(PVR_CHANNEL_PROVIDER_TYPE_ADDON),
+    m_clientProviderType(PVR_CHANNEL_PROVIDER_TYPE_ADDON),
+    m_strIconPath(client->Icon()),
+    m_strClientIconPath(client->Icon()),
+    m_bIsClientProvider(true)
+{
+  auto& artMap = client->GetAddonInfo()->Art();
+  auto thumbEntry = artMap.find("thumb");
+  if (thumbEntry != artMap.end())
+    m_strThumbPath = thumbEntry->second;
+
+  UpdatePath();
+}
+
+bool CPVRChannelProvider::operator==(const CPVRChannelProvider& right) const
+{
+  return (m_iUniqueId == right.m_iUniqueId && m_iClientId == right.m_iClientId);
+}
+
+/**
+ * Compare not equal for two CPVRChannelProvider
+ */
+bool CPVRChannelProvider::operator!=(const CPVRChannelProvider& right) const
+{
+  return !(*this == right);
+}
+
+void CPVRChannelProvider::Serialize(CVariant& value) const
+{
+  value["providerid"] = m_iProviderId;
+  value["clientid"] = m_iClientId;
+  value["providername"] = m_strProviderName;
+  switch (m_providerType)
+  {
+    case PVR_CHANNEL_PROVIDER_TYPE_ADDON:
+      value["providertype"] = "addon";
+      break;
+    case PVR_CHANNEL_PROVIDER_TYPE_SATELLITE:
+      value["providertype"] = "satellite";
+      break;
+    case PVR_CHANNEL_PROVIDER_TYPE_CABLE:
+      value["providertype"] = "cable";
+      break;
+    case PVR_CHANNEL_PROVIDER_TYPE_AERIAL:
+      value["providertype"] = "aerial";
+      break;
+    case PVR_CHANNEL_PROVIDER_TYPE_IPTV:
+      value["providertype"] = "iptv";
+      break;
+    default:
+      value["state"] = "unknown";
```

Comment:
Why do you serialize the type to strings? Couldn't this simply be the enum's numerical value?

Reply:
Because it will be used in the JSON API (eventually) and numbers don't mean much. On a previous json PR this was pointed out so I doing it here too.

Reply:
Okay, cool.